File: cpp/EndpointBase.cpp
/*
 * EndpointBase.cpp
 *
 *  Created on: Nov 28, 2021
 *      Author: mballance
 */

#include <stdarg.h>
#include "Debug.h"
#include "EndpointBase.h"
#include "EndpointEventBase.h"
#include "EndpointListenerBase.h"

#include "InterfaceInstBase.h"
#include "InterfaceType.h"
#include "InterfaceTypeBuilder.h"
#include "ParamValBool.h"
#include "ParamValInt.h"
#include "ParamValMap.h"
#include "ParamValStr.h"
#include "ParamValVec.h"

#define EN_DEBUG_ENDPOINT_BASE

#ifdef EN_DEBUG_ENDPOINT_BASE
#define DEBUG_ENTER(fmt, ...) \
	DEBUG_ENTER_BASE(EndpointBase, fmt, ##__VA_ARGS__)
#define DEBUG_LEAVE(fmt, ...) \
	DEBUG_LEAVE_BASE(EndpointBase, fmt, ##__VA_ARGS__)
#define DEBUG(fmt, ...) \
	DEBUG_BASE(EndpointBase, fmt, ##__VA_ARGS__)
#else
#define DEBUG_ENTER(fmt, ...)
#define DEBUG_LEAVE(fmt, ...)
#define DEBUG(fmt, ...)
#endif

namespace tblink_rpc_core {

EndpointBase::EndpointBase() : m_flags(IEndpointFlags::Empty) {
	m_time = 0;
	m_time_precision = 0;

	m_init = 0;
	m_peer_init = 0;
	m_build_complete = 0;
	m_peer_build_complete = 0;
	m_build_check_complete = 0;
	m_connect_complete = 0;
	m_peer_connect_complete = 0;
	m_connect_check_complete = 0;

	m_comm_state = IEndpoint::Waiting;
	m_comm_mode = IEndpoint::Automatic;

	m_ifinst_id = 0;
}

EndpointBase::~EndpointBase() {
	// TODO Auto-generated destructor stub
}

IEndpointFlags EndpointBase::getFlags() {
	return m_flags;
}

void EndpointBase::setFlag(IEndpointFlags f) {
	m_flags |= f;
}

int32_t EndpointBase::init(IEndpointServices *services) {
	m_services = IEndpointServicesUP(services);

	if (m_services) {
		m_services->init(this);
	}

	return 0;
}

int32_t EndpointBase::is_init() {
	DEBUG("is_init m_init=%d m_peer_init=%d", m_init, m_peer_init);
	if (m_init == -1 || m_peer_init == -1) {
		return -1;
	} else {
		return m_init && m_peer_init;
	}
}

IEndpoint::comm_state_e EndpointBase::comm_state() {
	return m_comm_state;
}

void EndpointBase::update_comm_mode(
		IEndpoint::comm_mode_e m,
		IEndpoint::comm_state_e s) {
	DEBUG_ENTER("update_comm_mode");

	m_comm_mode = m;
	m_comm_state = s;

	DEBUG_LEAVE("update_comm_mode");
}

int32_t EndpointBase::build_complete() {
	m_build_complete = 1;
	return 0;
}

int32_t EndpointBase::is_build_complete() {
	DEBUG("is_build_complete local=%d peer=%d",
			m_build_complete, m_peer_build_complete);
	if (!m_build_check_complete) {
		if (m_build_complete && m_peer_build_complete) {
			m_build_check_complete = build_complete_check();
		}
	}
	return m_build_check_complete;
}

int32_t EndpointBase::connect_complete() {
	m_connect_complete = 1;

	if (m_peer_connect_complete) {
		m_connect_check_complete = connect_complete_check();
	}
	return 0;
}

int32_t EndpointBase::is_connect_complete() {
	if (!m_connect_check_complete) {
		if (m_connect_complete && m_peer_connect_complete) {
			m_connect_check_complete = connect_complete_check();
		}
	}
	return m_connect_check_complete;
}

IEndpointListener *EndpointBase::addListener(const endpoint_ev_f &ev_f) {
	DEBUG_ENTER("addListener");
	EndpointListenerBase *l = new EndpointListenerBase(ev_f);
	m_listeners.push_back(IEndpointListenerUP(l));
	m_listeners_p.push_back(l);
	DEBUG_LEAVE("addListener %d", m_listeners_p.size());
	return l;
}

void EndpointBase::addListener(IEndpointListener *l) {
	DEBUG_ENTER("addListener");
	m_listeners_p.push_back(l);
	DEBUG_LEAVE("addListener %d", m_listeners_p.size());
}

void EndpointBase::removeListener(IEndpointListener *l) {
	for (auto it=m_listeners.begin(); it!=m_listeners.end(); it++) {
		if (it->get() == l) {
			m_listeners.erase(it);
			break;
		}
	}
	for (auto it=m_listeners_p.begin(); it!=m_listeners_p.end(); it++) {
		if (*it == l) {
			m_listeners_p.erase(it);
			break;
		}
	}
}

void EndpointBase::sendEvent(IEndpointEvent::kind_t kind) {
	EndpointEventBase ev(kind);
	sendEvent(&ev);
}

void EndpointBase::sendEvent(const IEndpointEvent *ev) {
	DEBUG_ENTER("sendEvent: %d", ev->kind());
	for (int32_t i=0; i<m_listeners_p.size(); i++) {
		IEndpointListener *l = m_listeners_p.at(i);
		DEBUG_ENTER("--> Send to %d", i);
		l->event(ev);
		DEBUG_ENTER("<-- Send to %d", i);
		if (i < m_listeners_p.size() && m_listeners_p.at(i) != l) {
			// list has changed
			i--;
		}
	}
	DEBUG_LEAVE("sendEvent: %d", ev->kind());
}

std::vector<std::string> EndpointBase::args() {
	if (m_services) {
		return m_services->args();
	} else {
		// TODO:
		return m_args;
	}
}

uint64_t EndpointBase::time() {
	if (m_services) {
		return m_services->time();
	} else {
		return m_time;
	}
}

int32_t EndpointBase::time_precision() {
	if (m_services) {
		return m_services->time_precision();
	} else {
		return m_time_precision;
	}
}

IInterfaceType *EndpointBase::findInterfaceType(
		const std::string		&name) {
	auto it = m_local_ifc_types.find(name);

	if (it != m_local_ifc_types.end()) {
		return it->second;
	} else {
		return 0;
	}
}

IInterfaceTypeBuilder *EndpointBase::newInterfaceTypeBuilder(
		const std::string		&name) {
	return new InterfaceTypeBuilder(name);
}

IInterfaceType *EndpointBase::defineInterfaceType(
		IInterfaceTypeBuilder	*type,
		IInterfaceImplFactory	*impl_factory,
		IInterfaceImplFactory	*impl_mirror_factory) {
	InterfaceTypeBuilder *builder = static_cast<InterfaceTypeBuilder *>(type);
	InterfaceType *iftype = builder->type();

	iftype->impl_f(impl_factory);
	iftype->impl_mirror_f(impl_mirror_factory);








	auto it = m_local_ifc_types.find(iftype->name());
	if (it != m_local_ifc_types.end()) {
		DEBUG("IfType %s exists", iftype->name().c_str());
	}

	DEBUG_ENTER("defineInterfaceType %s", iftype->name().c_str());
	m_local_ifc_types.insert({iftype->name(), iftype});
	m_local_ifc_type_pl.push_back(iftype);
	m_local_ifc_type_l.push_back(InterfaceTypeUP(iftype));

	DEBUG_LEAVE("defineInterfaceType %s", iftype->name().c_str());
	return iftype;
}

const std::vector<IInterfaceType *> &EndpointBase::getInterfaceTypes() {
	return m_local_ifc_type_pl;
}

const std::vector<IInterfaceType *> &EndpointBase::getPeerInterfaceTypes() {
	return m_peer_ifc_types_pl;
}

const std::vector<IInterfaceInst *> &EndpointBase::getInterfaceInsts() {
	return m_local_ifc_insts_pl;
}

const std::vector<IInterfaceInst *> &EndpointBase::getPeerInterfaceInsts() {
	return m_peer_ifc_insts_pl;
}

IParamValBool *EndpointBase::mkValBool(bool val) {
	return new ParamValBool();
}

IParamValInt *EndpointBase::mkValIntU(uint64_t val, int32_t width) {
	return new ParamValInt(val);
}

IParamValInt *EndpointBase::mkValIntS(int64_t val, int32_t width) {
	return new ParamValInt(val);
}

IParamValMap *EndpointBase::mkValMap() {
	return new ParamValMap();
}

IParamValStr *EndpointBase::mkValStr(const std::string &val) {
	return new ParamValStr(val);
}

IParamValVec *EndpointBase::mkValVec() {
	return new ParamValVec();
}

int EndpointBase::peer_init(
		int32_t								time_precision,
		const std::vector<std::string>		&args) {
	m_peer_init = 1;
	m_time_precision = time_precision;
	fprintf(stdout, "peer_init:\n");
	for (auto it=args.begin(); it!=args.end(); it++) {
		fprintf(stdout, "    arg: %s\n", it->c_str());
		m_args.push_back(*it);
	}

	return 0;
}

int EndpointBase::peer_build_complete() {
	m_peer_build_complete = 1;

	return 0;
}

int32_t EndpointBase::build_complete_check() {
	int32_t ret = 1;

	// Ensure that both sides have the same types?
	return 1;
}

int EndpointBase::peer_connect_complete() {
	m_peer_connect_complete = 1;
	return 0;
}

int32_t EndpointBase::connect_complete_check() {
	int32_t ret = 1;
	DEBUG_ENTER("connect_complete_check");

	// Check results
	for (auto it=m_local_ifc_insts.begin();
			it!=m_local_ifc_insts.end(); it++) {
		ifinst_m_t::iterator peer_it;
		if ((peer_it=m_peer_ifc_insts.find(it->first)) != m_peer_ifc_insts.end()) {
			// Compare types
			IInterfaceInst *local_ifinst = it->second;
			IInterfaceInst *peer_ifinst = peer_it->second;

			if (local_ifinst->type()->name() != peer_ifinst->type()->name()) {
				fprintf(stdout, "local_ifinst::type=%p peer_ifinst::type=%p\n",
						local_ifinst->type(), peer_ifinst->type());
				fflush(stdout);
				last_error("Local interface %s is of type %s ; Peer is of type %s\n",
						local_ifinst->name().c_str(),
						local_ifinst->type()->name().c_str(),
						peer_ifinst->type()->name().c_str());
				ret = -1;
				break;
			}
			// Configure the local interface so it knows how to
			// talk to the peer.
			dynamic_cast<InterfaceInstBase *>(local_ifinst)->setRemoteId(
					dynamic_cast<InterfaceInstBase *>(peer_ifinst)->getLocalId());
		} else {
			// Instance doesn't exist
			last_error("Local interface %s does not have remote peer\n",
					it->second->name().c_str());
			ret = -1;
			break;
		}
	}

	// Check the opposite way
	if (ret == 1) {
		for (auto it=m_peer_ifc_insts.begin();
				it!=m_peer_ifc_insts.end(); it++) {
			ifinst_m_t::iterator local_it;
			if ((local_it=m_local_ifc_insts.find(it->first)) != m_local_ifc_insts.end()) {
				// Compare types
				IInterfaceInst *peer_ifinst = it->second;
				IInterfaceInst *local_ifinst = local_it->second;

				if (local_ifinst->type()->name() != peer_ifinst->type()->name()) {
					fprintf(stdout, "Error: Local interface %s is of type %s ; Peer is of type %s\n",
							local_ifinst->name().c_str(),
							local_ifinst->type()->name().c_str(),
							peer_ifinst->type()->name().c_str());
					ret = -1;
					break;
				}
			} else {
				// Instance doesn't exist
				fprintf(stdout, "Error: Local interface %s does not have remote peer\n",
						it->second->name().c_str());
				ret = -1;
				break;
			}
		}
	}

//	if (ret == 1) {
//		// Populate the local id->inst map
//		for (auto it=m_local_ifc_insts.begin();
//				it!=m_local_ifc_insts.end(); it++) {
//			InterfaceInstBase *ifinst =
//					dynamic_cast<InterfaceInstBase *>(it->second.get());
//			DEBUG("Add interface-id %lld", ifinst->getLocalId());
//			m_id2ifinst_m.insert({ifinst->getLocalId(), ifinst});
//		}
//	}

	DEBUG_LEAVE("connect_complete_check %d", ret);

	return ret;
}

void EndpointBase::last_error(const char *fmt, ...) {
	va_list ap;
	char tmp[256];

	va_start(ap, fmt);
	vsnprintf(tmp, sizeof(tmp), fmt, ap);
	m_last_error = fmt;
	va_end(ap);
}

} /* namespace tblink_rpc_core */

File: python/tblink_rpc_core/endpoint_base.py
'''
Created on Feb 28, 2022

@author: mballance
'''
from tblink_rpc_core.endpoint import Endpoint, EndpointFlags, comm_mode_e, \
    comm_state_e
from tblink_rpc_core.endpoint_event import EndpointEvent
from tblink_rpc_core.endpoint_listener import EndpointListener
from tblink_rpc_core.event_type_e import EventTypeE
from tblink_rpc_core.interface_impl_factory import InterfaceImplFactory
from tblink_rpc_core.interface_type import InterfaceType
from tblink_rpc_core.interface_type_builder import InterfaceTypeBuilder
from tblink_rpc_core.param_val_bool import ParamValBool
from tblink_rpc_core.param_val_int import ParamValInt
from tblink_rpc_core.param_val_map import ParamValMap
from tblink_rpc_core.param_val_str import ParamValStr
from tblink_rpc_core.param_val_vec import ParamValVec


class EndpointBase(Endpoint):
    
    def __init__(self):
        self._flags = EndpointFlags.Empty
        self._comm_mode = comm_mode_e.Automatic
        self._comm_state = comm_state_e.Waiting
        self._eps = None
        self._endpoint_listener_m = {}
        self._local_interface_type_l = []
        self._local_interface_type_m = {}
        self._peer_interface_type_l = []
        self._peer_interface_type_m = {}
        self._local_interface_inst_l = []
        self._local_interface_inst_m = {}
        self._peer_interface_inst_l = []
        self._peer_interface_inst_m = {}

    def getFlags(self) -> EndpointFlags:
        return self._flags
        
    def setFlag(self, f) -> EndpointFlags:
        self._flags |= f

    def init(self, ep_services : 'EndpointServices') -> bool:
        self._eps = ep_services
        pass
    
    def is_init(self) -> int:
        """
        Returns True when init has been signaled by this EP,
        and init has been signaled by the peer EP
        """
        raise NotImplementedError("is_init for class %s" % str(type(self)))
    
    def build_complete(self) -> bool:
        """
        Notifies the peer EP that build activities are complete
        """
        raise NotImplementedError("build_complete for class %s" % str(type(self)))
            
    def is_build_complete(self) -> int:
        """
        Returns True when build_complete has been signaled by this EP,
        and build_complete has been signaled by the peer EP
        """
        raise NotImplementedError("is_build_complete for class %s" % str(type(self)))

    def connect_complete(self) -> bool:
        """
        Notifies the peer EP that connect activities are complete
        """
        raise NotImplementedError("connect_complete for class %s" % str(type(self)))
            
    def is_connect_complete(self) -> int:
        """
        Returns True when connect_complete has been signaled by this EP,
        and connect_complete has been signaled by the peer EP
        """
        raise NotImplementedError("is_connect_complete for class %s" % str(type(self)))

    def addListener(self, listener_f) -> EndpointListener:
        self._endpoint_listener_m[listener_f] = listener_f
        return listener_f
    
    def removeListener(self, listener_h):
        if listener_h in self._endpoint_listener_m.keys():
            self._endpoint_listener_m.pop(listener_h)
            
    def comm_state(self) -> comm_state_e:
        return self._comm_state

    def update_comm_mode(self, m : comm_mode_e, s : comm_state_e):
        self._comm_mode = m 
        self._comm_state = s
        
    def findInterfaceType(self, name) -> InterfaceType:
        if name in self._local_interface_type_m.keys():
            return self._local_interface_type_m[name]
        else:
            return None

    def newInterfaceTypeBuilder(self, name) -> InterfaceTypeBuilder:
        return InterfaceTypeBuilder(name)
    
    def defineInterfaceType(self, 
                            iftype_b : InterfaceTypeBuilder,
                            impl_f : InterfaceImplFactory,
                            impl_mirror_f : InterfaceImplFactory) -> InterfaceType:
        t = InterfaceType(iftype_b.name)
        t._method_m.update(iftype_b.method_m)
        t._methods.extend(iftype_b.methods)
        
        self._local_interface_type_l.append(t)
        self._local_interface_type_m[iftype_b.name] = t
        
        return t

    def getInterfaceTypes(self):
        return self._local_interface_type_l
    
    def getPeerInterfaceTypes(self):
        return self._peer_interface_type_l
    
    def getInterfaceInsts(self):
        return self._local_interface_inst_l
    
    def getPeerInterfaceInsts(self):
        return self._peer_interface_inst_l 

    def shutdown(self):
        """
        Sends a shutdown-request message to the peer EP
        """
        print("Sending shutdown event")
        self._sendEvent(EndpointEvent(EventTypeE.Terminate))


    def _sendEvent(self, ev):
        items = list(self._endpoint_listener_m.items())
        
        for it in items:
            it[1](ev)
            
    def mkValBool(self, v) -> ParamValBool:
        return ParamValBool(v)
    
    def mkValIntS(self, val, width) -> ParamValInt:
        return ParamValInt(val)
    
    def mkValIntU(self, val, width) -> ParamValInt:
        return ParamValInt(val)
    
    def mkValMap(self) -> ParamValMap:
        return ParamValMap()
    
    def mkValVec(self) -> ParamValVec:
        return ParamValVec()
    
    def mkValStr(self, v) -> ParamValStr:
        return ParamValStr(v)
